python/dgl/tvm/gspmm.py

```diff
@@ -0,0 +1,282 @@
+    
+def _spmm_cuda_general(s, out):
+    edge_axis = out.op.axis[0]
+    feat_axis = s[out].fuse(*out.op.axis[1:])
+    s[out].bind(edge_axis, te.thread_axis('blockIdx.x'))
+    s[out].bind(feat_axis, te.thread_axis('threadIdx.x'))
+
+def _spmm_cuda_tree_reduce(s, out):
+    reduce_axis = out.op.reduce_axis[0]
+    edge_axis = out.op.axis[0]
+    feat_axis = s[out].fuse(*out.op.axis[1:])
+    ro, ri = s[out].split(reduce_axis, factor=32)
+    s[out].bind(ri, te.thread_axis('threadIdx.x'))
+    s[out].bind(feat_axis, te.thread_axis('threadIdx.y'))
+    s[out].bind(edge_axis, te.thread_axis('blockIdx.x'))
+
+def _spmm_cpu(s, out, num_feat_partitions, inlines, reshapes):
+    reduce_axis = out.op.reduce_axis[0]
+    row_axis = out.op.axis[0]
+    if num_feat_partitions == 1:
+        s[out].reorder(row_axis, reduce_axis, *out.op.axis[1:])
+    else:
+        feat_axis = s[out].fuse(*out.op.axis[1:])
+        fo, fi = s[out].split(feat_axis, nparts=num_feat_partitions)
+        s[out].reorder(fo, row_axis, reduce_axis, fi)
+        for t in inlines:
+            s[t].compute_inline()
+        for rs in reshapes:
+            s[rs].compute_at(s[out], fo)
+            s[rs].parallel(rs.op.axis[1])
+    s[out].parallel(row_axis)
+
+def _spmm_dds_sched(s, out, I, num_feat_partitions, inlines, reshapes):
+    ifeat_axis = s[I].fuse(*I.op.axis[2:])
+    s[I].reorder(I.op.axis[0], I.op.axis[1], I.op.reduce_axis[0], ifeat_axis)
+    ofeat_axis = s[out].fuse(*out.op.axis[1:])
+    if num_feat_partitions == 1:
+        s[out].reorder(out.op.reduce_axis[0], out.op.axis[0], ofeat_axis)
+    else:
+        # only need to partition feature axis of out
+        ofo, ofi = s[out].split(ofeat_axis, nparts=num_feat_partitions)
+        s[out].reorder(ofo, out.op.reduce_axis[0], out.op.axis[0], ofi)
+        for t in inlines:
+            s[t].compute_inline()
+        for t in reshapes:
+            s[t].compute_at(s[out], ofo)
+            s[t].parallel(t.op.axis[1])
+    s[I].compute_at(s[out], out.op.reduce_axis[0])
+    s[I].parallel(I.op.axis[1])
+    # s[I].vectorize(I.op.axis[2])
+    s[out].parallel(out.op.axis[0])
+
+def spmm(binary_op, reduce_op, nnz, num_rows, num_cols, 
+         lhs_shp, rhs_shp, out_shp,
+         indice_type, feat_type, use_idx=False,
+         num_col_partitions=1, num_feat_partitions=1, target='llvm'):
+    if target == 'cuda' and (num_col_partitions > 1 or num_feat_partitions > 1):
+        raise NotImplementedError
+    if '32' in indice_type:
+        indice_type = 'int32'
+    elif '64' in indice_type:
+        indice_type = 'int64'
+    else:
+        raise NotImplementedError
+    if '32' in feat_type:
+        feat_type = 'float32'
+    elif '64' in feat_type:
+        feat_type = 'float64'
+    elif '16' in feat_type:
+        feat_type = 'float16'
+    else:
+        raise NotImplementedError
+    # check if used for sampling
+    generic_shape = nnz == 0 and num_rows == 0 and num_cols == 0
```

Comment:
If user would like to deal with generic shape, `nnz`, `num_rows` and `num_cols` need to be set to all zeros?

Reply:
Right. We can use another flag, if you prefer.